src/pymatgen/analysis/pourbaix_diagram.py

```diff
@@ -377,34 +395,30 @@ def from_dict(cls, dct: dict) -> Self:
             dct.get("attribute"),
         )
 
-    def as_dict(self):
+    def as_dict(self) -> dict[Literal["ion", "energy", "name"], Any]:
         """Create a dict of composition, energy, and ion name."""
         return {"ion": self.ion.as_dict(), "energy": self.energy, "name": self.name}
 
-    def __repr__(self):
-        return f"IonEntry : {self.composition} with energy = {self.energy:.4f}"
-
 
-def ion_or_solid_comp_object(formula):
+def ion_or_solid_comp_object(formula: str) -> Composition | Ion:
     """Get an Ion or Composition object given a formula.
 
     Args:
-        formula: String formula. Eg. of ion: NaOH(aq), Na[+];
-            Eg. of solid: Fe2O3(s), Fe(s), Na2O
+        formula (str): Formula. E.g. of ion: NaOH(aq), Na[+];
+            E.g. of solid: Fe2O3(s), Fe(s), Na2O
 
     Returns:
         Composition/Ion object
     """
-    if re.match(r"\[([^\[\]]+)\]|\(aq\)", formula):
-        comp_obj = Ion.from_formula(formula)
-    elif re.search(r"\(s\)", formula):
-        comp_obj = Composition(formula[:-3])
-    else:
-        comp_obj = Composition(formula)
-    return comp_obj
+    # Formula for ion
+    try:
+        return Ion.from_formula(formula)
```

Comment:
TODO: one issue with current implementation, `Na2O` would be treated as `Ion` but from the docstring it should be `Composition`